src/nupic/engine/Link.cpp

```diff
@@ -396,17 +407,164 @@ void Link::write(LinkProto::Builder& proto) const
   proto.setSrcOutput(srcOutputName_.c_str());
   proto.setDestRegion(destRegionName_.c_str());
   proto.setDestInput(destInputName_.c_str());
+
+  // Save dealayed outputs
+  auto delayedOutputsBuilder = proto.initDelayedOutputs(propagationDelay_);
+  for (size_t i=0; i < propagationDelay_; ++i)
+  {
+    const Array & array = srcBuffer_[i];
+
+    auto genericArrayBuilder = delayedOutputsBuilder[i];
+
+    const size_t elementCount = array.getCount();
+
+    const auto arrayType = array.getType();
+
+    switch (arrayType)
```

Comment:
`copyArrayToArrayProto` won't need to be templated anymore since we can determine appropriate types inside the switch/case blocks. If you don't want to repeat the for loop inside every `case` then you can have a generic `copyArrayToArrayProto` with the switch statement and then a simple `_tempaltedCopyArrayToArrayProto` like the current one.

The main point is that we don't want the big `switch` statement in the client code - we want it in the utility function for arrays that can be reused in different places.